### PR TITLE
Add ability to customize docker image using labels

### PR DESCRIPTION
**Pull request by itanf**
Implemented option from comment to PR https://github.com/jenkinsci/mesos-plugin/pull/198

### CODE REVIEW

**Comment by carlossg** [comment]
FWIW you can update the existing PR by pushing to the same branch, instead of creating a new PR

**Comment by itanf** [comment]
@Vlatombe, could you please review my PR

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 1]
```diff
@@ -170,39 +174,105 @@ public MesosSlaveInfo(
...
+              slaveAttributes,
+              jvmArgs,
+              jnlpArgs,
+              defaultSlave,
+              containerInfo.copyWithDockerImage(dockerImage),
+              additionalURIs,
+              nodeProperties
+      );
+    } catch (Descriptor.FormException e) {
+      return null;
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Should log the exception.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 2]
```diff
@@ -170,39 +174,105 @@ public MesosSlaveInfo(
...
+    } catch (IOException e) {
+      return null;
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Should log the exception.

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosSlaveInfo/ContainerInfo/help-dockerImageCustomizable.html [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<div>
+    This will allow override default docker image using labels. Example if mesos slave has lable mesosSlave and you
```
Comment (Vlatombe) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosSlaveInfo/ContainerInfo/help-dockerImageCustomizable.html]: typo on 'lable'

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 4]
```diff
@@ -308,6 +390,42 @@ public Object readResolve() {
     return this;
   }
 
+  private static String getCustomImage(Label label) {
+    Pattern r = Pattern.compile(CUSTOM_IMAGE_SEPARATOR + "([\\w\\.\\-/:]+[\\w])");
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Should be a constant.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 5]
```diff
@@ -308,6 +390,42 @@ public Object readResolve() {
...
+    }
+
+    return null;
+  }
+
+  private static Label getLabelWithoutCustomImage(Label label, String customDockerImage) {
+    return Label.get(label.toString().replace(CUSTOM_IMAGE_SEPARATOR + customDockerImage, ""));
+  }
+
+  public MesosSlaveInfo getMesosSlaveInfoForLabel(Label label) {
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: This deserves a unit test.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 6]
```diff
@@ -297,8 +367,20 @@ private String cleanseJvmArgs(final String jvmArgs) {
    * @return Whether the slave label matches.
    */
   public boolean matchesLabel(@CheckForNull Label label) {
-    return ((label == null) && (getLabelString() == null))
-      || (getLabelString() != null && label != null && label.matches(Label.parse(getLabelString())));
+    if (label == null || getLabelString() == null) {
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Deserves a unit test.

**Comment by itanf** [comment]
@Vlatombe, Thanks for quick response!
I've implemented the requested changes so could you please take a look then you have time.

**Comment by Vlatombe** [comment]
I'm getting a NPE on the following assertion added to `getMesosSlaveInfoForLabelTest` 

    assertEquals(null, buildMesosSlaveInfo(null, false).getMesosSlaveInfoForLabel(null).getLabelString());

This matches a case where a slave configuration with a null (empty) labelString (cf. https://github.com/jenkinsci/mesos-plugin/pull/178 introducing this feature)

**Comment by itanf** [comment]
@Vlatombe, Thank you for noticed! I've added a few assertions and fix logic.

**Comment by Vlatombe** [comment]
LGTM. Thank you for your contribution!

**Pull request merged**